openeogeotrellis/job_tracker_v2.py

```diff
@@ -85,48 +87,112 @@ class YarnAppReportParseException(Exception):
-                application_report = subprocess.check_output(
-                    command, stderr=subprocess.PIPE
-                ).decode("utf-8")
-        except CalledProcessError as e:
-            stdout = e.stdout.decode()
-            if "doesn't exist in RM or Timeline Server" in stdout:
-                raise AppNotFound(stdout) from None
-            else:
-                raise
-        return self.parse_application_report(application_report)
+        # TODO: Are there any cases were we *do* need to provide credentials to the Kerberos authentication?
+        #   At present we are assuming there is already a Kerberos ticket because kinit was run before
+        #   the geopyspark driver process was started.
+        #   AFAIK that means we don't have to provide user & pw which is in fact the whole point.
+
+        url = self.get_application_url(application_id=app_id)
+        response = requests.get(url, auth=self.get_authentication_provider())
+
+        # Handle error responses: We need the field "message" from "RemoteException"
+        # See https://hadoop.apache.org/docs/current/hadoop-yarn/hadoop-yarn-site/WebServicesIntro.html
+        #   specifically, the section: Response Errors
+        #
+        # Example of an error responses body:
+        # {
+        #     "RemoteException": {
+        #         "exception": "NotFoundException",
+        #         "message": "java.lang.Exception: app with id: application_1674538064532_46723 not found",
+        #         "javaClassName": "org.apache.hadoop.yarn.webapp.NotFoundException"
+        #     }
+        # }
+        if response.status_code == 404:
+            remote_exc = response.json().get("RemoteException", {})
+            message = remote_exc.get(
+                "message",
+                "Unknown error. Error response did not contain the message property.",
+            )
+            raise AppNotFound(message) from None
+        else:
+            # Check if there was any other HTTP error status.
+            response.raise_for_status()
+            return self.parse_application_response(json=response.json())
+
+    @staticmethod
+    def _ms_epoch_to_date(epoch_millis: int) -> Union[str, None]:
+        """Parse millisecond timestamp from app report and return as rfc3339 date (or None)"""
+        if epoch_millis == 0:
+            return None
+        utc_datetime = dt.datetime.utcfromtimestamp(epoch_millis / 1000)
+        return rfc3339.datetime(utc_datetime)
+
+    @classmethod
+    def parse_application_response(cls, json: dict) -> _JobMetadata:
+        if not json:
+            raise YarnAppReportParseException("Response is empty")
+
+        # Handle a corrupt response that comes through as a string, not valid JSON.
+        if isinstance(json, str):
+            raise YarnAppReportParseException(
+                f"Response is corrupt, not valid JSON. Received following string: {json!r}"
+            )
```

Comment:
I would just assume that the response is a JSON object that will get parsed to a dict. `response.json()` will have raised an exception for any unexpected response anyway.

Reply:
This was bothering me a bit too.

But while trying to eliminate this check, I realized that a single string is apparently also valid JSON and because of that  response.json() doesn't raise an exception.

Also there was an existing test case for a corrupted response that just some piece of text. Though that was originally a test case for corrupted output of the yarn CLI command, and perhaps that test case is not truly useful anymore.

So I took a view and clarified that the expected type for the parameter `json` can be dict or str, and made the error message a bit more clear.

Reply:
I'm not sure how response.json() behaves but I should hope that it raises an exception on strings not enclosed in double quotes.

Reply:
As we discussed this morning, I will move this check one layer up to the calling method.
That is a more appropriate place to validate this, should we ever get incorrect responses.


I also dug a little bit deeper into this because I wanted to know why a response body that contained - _seemingly_ - invalid JSON was slipping through the cracks. 

### In short

The problem was in the mock, not in requests.json()

Initially, I was surprised when I saw that requests.Response.json() did not raise an exception when the response was seemingly a string without double quotes around it. But the unit test was not set up correctly to test that.

The mock produced different data than I thought it would, and it actually gave requests a string ***inside double quotes***. So requests received valid JSON and there is no lurking problem with requests.json() after all.

What is more, requests.json() uses json.loads(), so it does behave the same as json.loads(), as we would expect.


For some reason the mock always adds double quotes around the response text if the response body is just a string, probably because we use in a way that is meant to produce JSON responses specifically.

If we wanted to simulate this particular (edge) case, where requests should detect invalid JSON, I think we would need to use requests_mock in a different way and create a raw text response. But that is a bit more cumbersome, and I think I will leave it at that.

Reply:
> One more thing that has to be verified in the context of replacing the YARN CLI with `requests`: the YARN resource manager runs in an HA setup so you have two nodes: https://epod-master1.vgt.vito.be:8090 and https://epod-master2.vgt.vito.be:8090, where one is active and the other one is passive.
> 
> If you send a request to epod-master1.vgt.vito.be and it is the active instance, it will Just Work. If, on the other hand, epod-master1.vgt.vito.be is currently passive, it will respond with a redirect to active node epod-master2.vgt.vito.be and you have to make sure that `requests` handles this redirect.

The requests library should already handle this. 

In the API docs we find that `requests.get(...) requests.post(...)` etc have the parameter  
`allow_redirects`  and that it defaults to True

See also: https://requests.readthedocs.io/en/latest/api/

Reply:
Turns out it wasn't that complicated to improve the unit test after all.

So now we cover both cases: 
- Invalid JSON gets caught by responses.json() and raises a JSONDecodeError as expected.
- If the response contains valid JSON that is not a dict, that will be caught by our own code and it raises YarnAppReportParseException